pennylane/debugging/operation_checker.py

```diff
@@ -0,0 +1,1095 @@
+    MatrixUndefinedError,
+    SparseMatrixUndefinedError,
+    GeneratorUndefinedError,
+    DiagGatesUndefinedError,
+    EigvalsUndefinedError,
+    TermsUndefinedError,
+    DecompositionUndefinedError,
+    AnyWires,
+)
+
+try:
+    import jax
+
+    has_jax = True
+except ModuleNotFoundError:
+    has_jax = False
+try:
+    import tensorflow as tf
+
+    has_tf = True
+except ModuleNotFoundError:
+    has_tf = False
+try:
+    import torch
+
+    has_torch = True
+except ModuleNotFoundError:
+    has_torch = False
+
+_colors = {
+    "error": "91",  # red
+    "fatal_error": "91",  # red
+    "hint": "93",  # yellow
+    "comment": 94,  # blue
+    "pass": "92",  # green
+}
+
+verbosity_levels = {"fatal_error": 0, "error": 1, "hint": 2, "comment": 3, "pass": 4}
+levels_verbosity = {val: key for key, val in verbosity_levels.items()}
+
+_default_methods_to_check = [
+    ("compute_eigvals", EigvalsUndefinedError, False),
+    ("compute_matrix", MatrixUndefinedError, False),
+    ("compute_sparse_matrix", SparseMatrixUndefinedError, False),
+    ("compute_terms", TermsUndefinedError, False),
+    ("compute_decomposition", DecompositionUndefinedError, True),
+    ("compute_diagonalizing_gates", DiagGatesUndefinedError, True),
+]
+
+
+def equal_up_to_phase(mat1, mat2, atol=1e-10):
+    r"""Check whether two matrices are equal up to a scalar
+    prefactor of the form :math:`\exp(i\phi)`.
+
+    Args:
+        mat1 (array_like): First matrix to check for equality
+        mat2 (array_like): Second matrix to check for equality
+        atol (float): Absolute tolerance for the check for equality
+
+    Return:
+        bool: Whether the two input matrices are equal up to a scalar
+        phase prefactor.
+    """
+    # Check whether the matrices are equal
+    if np.allclose(mat1, mat2, atol=atol, rtol=0.0):
+        return True
+
+    # Compute the potential scalar prefactor from the first nonzero entry of mat2
+    ids = np.where(np.round(mat2, 10))
+    idx = (ids[0][0], ids[1][0])
+    phase = mat1[idx] / mat2[idx]
+
+    # Return whether the matrices are equal, accounting for the potential scalar prefactor
+    return np.isclose(np.abs(phase), 1.0, atol=atol) and np.allclose(
+        mat1, mat2 * phase, atol=atol, rtol=0.0
+    )
+
+
+def is_diagonal(matrix, atol=1e-10):
+    r"""Check whether a matrix is a diagonal matrix
+
+    Args:
+        matrix (array_like): Matrix to check
+
+    Returns:
+        bool: Whether the input matrix is a diagonal matrix
+    """
+    # Extract the diagonal, subtract it from the input, and check whether the result is 0.
+    off_diagonal = matrix - np.diag(np.diag(matrix))
+    return np.allclose(off_diagonal, np.zeros_like(matrix), atol=atol, rtol=0.0)
+
+
+def torch_jacobian(fn):
+    """Functional jacobian in the torch interface with the same
+    syntax as qml.jacobian. Note that this is not performant!"""
+    return lambda *args: torch.autograd.functional.jacobian(fn, args)
```

Comment:
No idea what the functional jacobian is :)

Reply:
Oh, it returns a function that can be fed with parameters to construct the jacobian matrix?

Reply:
Ah, yes that is what I meant. But I suppose "with the same syntax as qml.jacobian" kind of contains that information. Will remove the "Functional"